### PR TITLE
Implement zlib compression

### PR DESCRIPTION
**Pull request by feliwir**
Use zlib library to optional de/compress data

### CODE REVIEW

File: src/game/common/compression/zlibcompr.cpp
Comment (xezon) [inline src/game/common/compression/zlibcompr.cpp]: eol

File: src/game/common/compression/zlibcompr.cpp
Comment (xezon) [inline src/game/common/compression/zlibcompr.cpp]: static cast, same in next function

File: src/game/common/compression/zlibcompr.h
Comment (xezon) [inline src/game/common/compression/zlibcompr.h]: eol

File: src/platform/audio/ffmpegaudiofilecache.cpp
Comment (xezon) [inline src/platform/audio/ffmpegaudiofilecache.cpp]: Why is this here?

File: src/game/common/compression/zlibcompr.cpp
Comment (xezon) [inline src/game/common/compression/zlibcompr.cpp]: Variable names need to be "name_name" format. Despite, these have the same name. Different name would be better.

File: src/game/common/compression/zlibcompr.cpp
Comment (xezon) [inline src/game/common/compression/zlibcompr.cpp]: <= 0 perhaps?

File: tests/test_compr.cpp
Comment (xezon) [inline tests/test_compr.cpp]: You can use `std::unique_ptr<>` to skip on `delete`. In several functions.

File: tests/test_compr.cpp
Comment (xezon) [inline tests/test_compr.cpp]: This compression result check looks quite generous, as it would allow error margin on compression result. We do expect compressed data to be always the same right? Perhaps hash contents of `dst_data` and compare hash value against expected hash value.

In multiple functions.

File: tests/test_compr.cpp
Comment (xezon) [inline tests/test_compr.cpp:70]: You can add a new line before to make this block a bit easier on the eyes.

File: tests/test_compr.cpp
Comment (xezon) [inline tests/test_compr.cpp]: I am not sure whether or not it is good to use `Utf8String` and `BufferedFileClass` in these compression test functions. It means these tests can fail if there is a problem in `Utf8String` & Co. If you intend to write separate tests for `Utf8String` and `BufferedFileClass`, then perhaps it would make sense to use `std::string` and `std::fstream` here instead. That way compression test is pure. Or we run tests for strings and files before this, so that they would fail before this.

File: src/game/common/compression/compressionmanager.h
Comment (xezon) [inline src/game/common/compression/compressionmanager.h]: Are these new functions as per original or are these custom for Thyme?

File: src/game/common/compression/compressionmanager.cpp
Comment (xezon) [inline src/game/common/compression/compressionmanager.cpp]: Omit the "\n" and "file a bug report".

In multiple log messages.

Also, perhaps it should be promoted to Assert? Could be separate change.

File: src/platform/audio/ffmpegaudiofilecache.cpp
Comment (feliwir) [reply src/platform/audio/ffmpegaudiofilecache.cpp]: Well i accidentally committed this since i need it locally to compile. However the diff should disappear once #780 is merged

File: tests/test_compr.cpp
Comment (feliwir) [reply tests/test_compr.cpp]: It's true that these function don't really check a lot, but i wanted to avoid that different zlib library version may fail. E.g. it's valid that a newer zlib version compresses better than an older one

File: src/game/common/compression/compressionmanager.h
Comment (feliwir) [reply src/game/common/compression/compressionmanager.h]: Apart from `Get_Compression_FourCC` all of these new functions exist in the original. Atleast that's the information @tomsons26  provided me with. I could probably get rid of `Get_Compression_FourCC` but i find it easier to read if there's a single place for that, instead of putting it everywhere

File: tests/test_compr.cpp
Comment (xezon) [reply tests/test_compr.cpp]: Hmm, is it true that compression result can differ that much, but then can still be read back with any zlib implementation? I would have expected that results will remain identical with identical settings, as otherwise it would not be able to decompress it again. I could be wrong.

Comment (feliwir) [reply tests/test_compr.cpp]: Well, depending on the compression level the size can vary from 728 bytes (level 1) -> 656 bytes (level 9) with different sizes inbetween accordingly. When the zlib maintainer finds a way to further decrease the size while maintaining compability with the DEFLATE algorithm he will certainly implement it. Also it might be possible that the flags for the different levels change and the size may change.

I know this is very unlikely, but the library doesn't give any guarantees that compression ratio doesn't change with different versions. E.g. if we wanted to switch to `libdeflate`in the future (a better zlib pretty much) we could get even better ratios: https://github.com/ebiggers/libdeflate#compression-levels

See here for a nice blog post: https://aras-p.info/blog/2021/08/09/EXR-libdeflate-is-great/

File: tests/test_compr.cpp
Comment (feliwir) [reply tests/test_compr.cpp]: Done

File: tests/test_compr.cpp
Comment (feliwir) [reply tests/test_compr.cpp]: Done

File: tests/test_compr.cpp
Comment (xezon) [reply tests/test_compr.cpp]: Is there a RefPack Decompress function? Perhaps pair that?

Comment (feliwir) [reply tests/test_compr.cpp]: What do you mean?

File: src/game/common/compression/compressionmanager.cpp
Comment (feliwir) [reply src/game/common/compression/compressionmanager.cpp]: I kept it as an error, changed the log message however

File: tests/test_compr.cpp
Comment (xezon) [reply tests/test_compr.cpp]: Original data -> RefPack Compress -> RefPack Decompress -> Compare decompressed with original.

File: src/game/common/compression/compressionmanager.cpp
Comment (xezon) [inline src/game/common/compression/compressionmanager.cpp]: I do not agree with this function contract. `const char*` should not be used for non-null terminated string buffer. Right now, function user makes assumption about buffer length (4). Imo this should return a struct with char[4] or a uint32_t. Also, we do actually have function and macros for proper FourCC. You can look around other code to see how it is used.

File: tests/test_compr.cpp
Comment (feliwir) [reply tests/test_compr.cpp]: I kind of do that with the current tests - i compress the files (and save them on disk) and then decompress those files on disk again. I don't compare the hashes, but atleast if the size matches.

However it's a good idea do make a roundtrip test, that does a `memcmp` on the decompressed data with the original

File: src/game/common/compression/compressionmanager.cpp
Comment (feliwir) [reply src/game/common/compression/compressionmanager.cpp]: Fixed

File: src/game/common/compression/compressionmanager.cpp
Comment (xezon) [inline src/game/common/compression/compressionmanager.cpp]: Spare semicolon.

File: tests/test_compr.cpp
Comment (xezon) [reply tests/test_compr.cpp]: Yes. I have seen the decompression in the other test. The simpler test would be to compress, compare, decompress, compare again, without writing compressed result to file. This reduces test file set. You would only need uncompr.txt. Or not even, you could make the text part of the cpp file and make everything in memory, skipping file IO entirely.

**Review by xezon — APPROVED** [review]
Merge #780 before this.

File: src/game/common/compression/compressionmanager.cpp
Comment (OmniBlade) [inline src/game/common/compression/compressionmanager.cpp:244]: Are we not missing a break here in case Zlib_Compress returns < 0?

**Review by OmniBlade — DISMISSED** [review]

File: src/game/common/compression/compressionmanager.cpp
Comment (feliwir) [reply src/game/common/compression/compressionmanager.cpp:244]: It can't return anything lower than 0

File: src/game/common/compression/zlibcompr.cpp
Comment (xezon) [inline src/game/common/compression/zlibcompr.cpp:54]: Optional: Log more detailed info about the type of error, and/or return different negative values depending on error type.

```
Z_BUF_ERROR
    The buffer dest was not large enough to hold the uncompressed data.

Z_MEM_ERROR
    Insufficient memory.

Z_DATA_ERROR
    The compressed data (referenced by source) was corrupted.
```

File: src/game/common/compression/zlibcompr.cpp
Comment (xezon) [inline src/game/common/compression/zlibcompr.cpp:69]: Optional: Log more detailed info about the type of error, and/or return different negative values depending on error type.

```
Z_BUF_ERROR
    The buffer dest was not large enough to hold the compressed data.

Z_MEM_ERROR
    Insufficient memory.

Z_STREAM_ERROR
    The level was not Z_DEFAULT_LEVEL, or was not between 0 and 9.
```

File: src/game/common/compression/compressionmanager.h
Comment (xezon) [reply src/game/common/compression/compressionmanager.h]: Could you perhaps add a comment in the header for `Get_Compression_FourCC` that it is Thyme specific?

**Pull request merged**
